diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7c2243b3b..e5ce85f55 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -313,6 +313,15 @@ set(BASE_SRCS
   src/type1/type1.c
   src/type42/type42.c
   src/winfonts/winfnt.c
+  src/otvalid/otvalid.c
+  src/otvalid/otvbase.c
+  src/otvalid/otvcommn.c
+  src/otvalid/otvgsub.c
+  src/otvalid/otvmod.c
+  src/otvalid/otvmath.c
+  src/otvalid/otvgpos.c
+  src/otvalid/otvgdef.c
+  src/otvalid/otvjstf.c
 )
 
 if (WIN32)
diff --git a/include/freetype/config/ftmodule.h b/include/freetype/config/ftmodule.h
index 76d271a74..0f4e18028 100644
--- a/include/freetype/config/ftmodule.h
+++ b/include/freetype/config/ftmodule.h
@@ -9,7 +9,7 @@
  *  FreeType without GNU make.
  *
  */
-
+FT_USE_MODULE( FT_Module_Class, otv_module_class )
 FT_USE_MODULE( FT_Module_Class, autofit_module_class )
 FT_USE_MODULE( FT_Driver_ClassRec, tt_driver_class )
 FT_USE_MODULE( FT_Driver_ClassRec, t1_driver_class )
diff --git a/src/otvalid/otvgsub.c b/src/otvalid/otvgsub.c
index c2b28569f..7d4c7bd8f 100644
--- a/src/otvalid/otvgsub.c
+++ b/src/otvalid/otvgsub.c
@@ -61,7 +61,8 @@
       {
         FT_Bytes  Coverage;
         FT_Int    DeltaGlyphID;
-        FT_Long   idx;
+        FT_UInt   first_cov, last_cov;
+        FT_UInt   first_idx, last_idx;
 
 
         OTV_LIMIT_CHECK( 4 );
@@ -70,12 +71,21 @@
 
         otv_Coverage_validate( Coverage, otvalid, -1 );
 
-        idx = (FT_Long)otv_Coverage_get_first( Coverage ) + DeltaGlyphID;
-        if ( idx < 0 )
+        first_cov = otv_Coverage_get_first( Coverage );
+        last_cov  = otv_Coverage_get_last( Coverage );
+
+        /* These additions are modulo 65536. */
+        first_idx = (FT_UInt)( (FT_Int)first_cov + DeltaGlyphID ) & 0xFFFFU;
+        last_idx  = (FT_UInt)( (FT_Int)last_cov + DeltaGlyphID ) & 0xFFFFU;
+
+        /* Since the maximum number of glyphs is 2^16 - 1 = 65535, */
+        /* the largest possible glyph index is 65534.  For this    */
+        /* reason there can't be a wrap-around region, which would */
+        /* imply the use of the invalid glyph index 65535.         */
+        if ( first_idx > last_idx )
           FT_INVALID_DATA;
 
-        idx = (FT_Long)otv_Coverage_get_last( Coverage ) + DeltaGlyphID;
-        if ( (FT_UInt)idx >= otvalid->glyph_count )
+        if ( last_idx >= otvalid->glyph_count )
           FT_INVALID_DATA;
       }
       break;
